model/riscv_vmem.sail

```diff
@@ -0,0 +1,855 @@
+       satp_asid_lsb_index = 44,
+
+       // SATP PPN
+       satp_ppn_size_bits  = 44,
+       satp_ppn_lsb_index  = 0,
+
+       // VA
+       va_size_bits        = 39,
+       vpn_size_bits       = 9,
+
+       // PTE
+       levels              = 3,
+       log_pte_size_bytes  = 3,    // 8 Bytes
+       pte_msbs_lsb_index  = 54,
+       pte_msbs_size_bits  = 10,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 44,
+       pte_PPN_j_size_bits = 9
+}
+
+// PRIVATE
+let sv48_params : SV_Params = struct {
+       // SATP CSR
+       satp_asid_size_bits = 16,
+       satp_asid_lsb_index = 44,
+
+       // SATP PPN
+       satp_ppn_size_bits  = 44,
+       satp_ppn_lsb_index  = 0,
+
+       // VA
+       va_size_bits        = 48,
+       vpn_size_bits       = 9,
+
+       // PTE
+       levels              = 4,
+       log_pte_size_bytes  = 3,    // 8 Bytes
+       pte_msbs_lsb_index  = 54,
+       pte_msbs_size_bits  = 10,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 44,
+       pte_PPN_j_size_bits = 9
+}
+
+// For future; not currently used
+// PRIVATE
+/*
+let sv57_params : SV_Params = struct {
+       // SATP CSR
+       satp_asid_size_bits = 16,
+       satp_asid_lsb_index = 44,
+
+       // SATP PPN
+       satp_ppn_size_bits  = 44,
+       satp_ppn_lsb_index  = 0,
+
+       // VA
+       va_size_bits        = 57,
+       vpn_size_bits       = 9,
+
+       // PTE
+       levels              = 5,
+       log_pte_size_bytes  = 3,    // 8 Bytes
+       pte_msbs_lsb_index  = 54,
+       pte_msbs_size_bits  = 10,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 44,
+       pte_PPN_j_size_bits = 9
+}
+*/
+
+// PRIVATE
+val      undefined_SV_Params : unit -> SV_Params
+
+function undefined_SV_Params() = sv32_params
+
+// ****************************************************************
+// Fields of virtual addresses
+
+// PRIVATE: Extract full VPN field from VA
+function vpns_of_va(sv_params : SV_Params,
+                    va        : bits(64)) -> bits(64) = {
+    let mask : bits(64) = (zero_extend(0b1) << sv_params.va_size_bits) - zero_extend(0b1);
+    (va & mask) >> PAGESIZE_BITS
+}
+
+// PRIVATE: Extract VPN[level] from  VA
+function vpn_j_of_va(sv_params : SV_Params,
+                     va        : bits(64),
+                     level     : nat)       -> bits(64) = {
+    let lsb  : nat       = PAGESIZE_BITS + level * sv_params.vpn_size_bits;
+    let mask : bits(64) = (zero_extend(0b1) << sv_params.vpn_size_bits) - 1;
+    ((va >> lsb) & mask)
+}
+
+// PRIVATE: Extract offset within page from VA
+function offset_of_va(va : bits(64)) -> bits(64) = {
+    let mask : bits(64) = (zero_extend(0b1) << PAGESIZE_BITS) - 1;
+    va & mask
+}
+
+// Valid xlen-wide values containing virtual addrs must have upper
+// bits equal to the MSB of the virtual address.
+// Virtual address widths depend on the virtual memory mode.
+// PRIVATE
+function is_valid_vAddr(struct { va_size_bits, _ } : SV_Params,
```

Comment:
Ok, thanks for the simpler way.

Re. “should it be xlenbits”: no, the current strategy for having a single piece of code working for Sv32, Sv39, Sv48 and Sv57 is to do all vmem work at 64 bits, parameterized by SV_Params for each specific case.  Note that even when in a system where xlen==64 (for M-mode), S-mode (SXL) and U-mode (UXL) may be 32 bit.

Reply:
I tried your suggested version of is_valid_vAddr().  It does not typecheck.

    Type error:
    model/riscv_vmem.sail:254.3-53:
    254 |   sign_extend(vAddr[va_size_bits - 1 .. 0]) == vAddr
        |   ^------------------------------------------------^
        | No overloading for (operator ==), tried:
        | * eq_int
        |    Cannot solve implicit 'm in sign_extend(subrange_bits(vAddr, sub_atom(va_size_bits, 1), 0))
        | * eq_bit
        |    Cannot solve implicit 'm in sign_extend(subrange_bits(vAddr, sub_atom(va_size_bits, 1), 0))
        | * eq_bool
        |    Cannot solve implicit 'm in sign_extend(subrange_bits(vAddr, sub_atom(va_size_bits, 1), 0))
        | * eq_unit
        |    Cannot solve implicit 'm in sign_extend(subrange_bits(vAddr, sub_atom(va_size_bits, 1), 0))
        | * eq_string
        |    Cannot solve implicit 'm in sign_extend(subrange_bits(vAddr, sub_atom(va_size_bits, 1), 0))
        | * eq_bit
        |    Cannot solve implicit 'm in sign_extend(subrange_bits(vAddr, sub_atom(va_size_bits, 1), 0))
        | * eq_bits
        |    Cannot solve implicit 'm in sign_extend(subrange_bits(vAddr, sub_atom(va_size_bits, 1), 0))
        | * eq_string
        |    Cannot solve implicit 'm in sign_extend(subrange_bits(vAddr, sub_atom(va_size_bits, 1), 0))
        | * eq_anything
        |    Cannot solve implicit 'm in sign_extend(subrange_bits(vAddr, sub_atom(va_size_bits, 1), 0))

Reply:
Could do `sign_extend(64, vAddr[va_size_bits - 1 .. 0]) == vAddr`

Reply:
`vAddr == sign_extend(vAddr[va_size_bits - 1 .. 0])` also works, as the type checker handles the arguments to `==` from left to right, and `vAddr` has the simpler type which then makes the sign_extend length obvious. I'm not sure I like that behavior but I don't know if there's any way to fix it.

Reply:
Ok, yes, this works.  Wow! I would not have suspected there’d be any difference just swapping the arg order.

I’ve made the change (in my clone; will push after some more consolidation).

Reply:
In this case I think it would be nice if the order didn't matter. I think what the type-checker would need to do would be to try to infer the first argument, then upon noticing it can't, continue checking the second argument. The second argument would then give us more information that would then 'invalidate' the failed check of the first argument, so we would re-check it. I'm not sure I'm quite up to implementing that right now though!